packages/aws-cdk-lib/aws-cloudwatch/lib/variable.ts

```diff
@@ -0,0 +1,171 @@
+   */
+  PROPERTY = 'property',
+
+  /**
+   * A pattern variable is one that changes a regex pattern across the dashboard JSON
+   */
+  PATTERN = 'pattern',
+}
+
+/**
+ * A single dashboard variable
+ */
+export interface IVariable {
+  /**
+   * Return the variable JSON for use in the dashboard
+   */
+  toJson(): any;
+}
+
+export interface IDashboardVariable {
+  /**
+   * Type of the variable
+   */
+  readonly type: VariableType;
+
+  /**
+   * The way the variable value is selected
+   */
+  readonly inputType: VariableInputType;
+
+  /**
+   * Pattern or property value to replace
+   */
+  readonly value: string;
+
+  /**
+   * Unique id
+   */
+  readonly id: string;
+
+  /**
+   * Optional label in the toolbar
+   */
+  readonly label?: string;
+
+  /**
+   * Optional default value
+   */
+  readonly defaultValue: any;
+
+  /**
+   * Whether the variable is visible
+   */
+  readonly visible?: boolean;
+}
+
+export abstract class DashboardVariable implements IVariable {
+  private readonly baseProps: IDashboardVariable;
+
+  protected constructor(props: IDashboardVariable) {
+    this.baseProps = props;
+  }
+
+  toJson(): any {
+    return {
+      [this.baseProps.type]: this.baseProps.value,
+      type: this.baseProps.type,
+      inputType: this.baseProps.inputType,
+      id: this.baseProps.id,
+      defaultValue: this.baseProps.defaultValue,
+      visible: this.baseProps.visible,
+      label: this.baseProps.label,
+    };
+  }
+
+}
+
+export interface IVariableValue {
+  /**
+   * Optional label for the selected item
+   *
+   * @default - value
+   */
+  readonly label?: string;
+  /**
+   * Value of the selected item
+   */
+  readonly value: string;
+}
+
+export interface IValueDashboardVariable extends IDashboardVariable {
+  /**
+   * List of custom values for the variable
+   *
+   * @default - No values
+   */
+  readonly values?: IVariableValue[];
+}
+
+/**
+ * The variable populated from the custom values
+ */
+export class ValueDashboardVariable extends DashboardVariable {
+  private readonly props: IValueDashboardVariable;
+
+  constructor(props: IValueDashboardVariable) {
+    super(props);
+    this.props = props;
+  }
+
+  toJson(): any {
+    const base = super.toJson();
+    return {
+      ...base,
+      values: this.props.values ? this.props.values.map(value => ({ label: value.label, value: value.value })) : undefined,
+    };
+  }
+}
+
+/**
+ * The variable populated from the metric search
+ */
+export interface ISearchDashboardVariable extends IDashboardVariable {
+  /**
+   * Search expression
+   */
+  readonly searchExpression: string;
+  /**
+   * Optional dimension name from the search
+   */
+  readonly populateFrom?: string;
+}
+
+export class SearchDashboardVariable extends DashboardVariable {
+  private readonly props: ISearchDashboardVariable;
+
+  constructor(props: ISearchDashboardVariable) {
```

Comment:
From the API docs, `search` and `populateFrom` are required  "if inputType is select or radio and you are not specifying values."

Just want to confirm that `SearchDashboardVariable` will still work with inputType `input` right?

Reply:
So, I went to try a different combo of parameters manually, and what would essentially happen if `input` inputType and we have a search variable, is that it will just be a normal `input`. However, on saving one is faced with the nice error
```
The dashboard body is invalid, there are 2 validation errors: [ { "dataPath": "/variables/0/search", "message": "Should NOT be valid" }, { "dataPath": "/variables/0/populateFrom", "message": "Should NOT be valid" } ]
```

So, to me there are 2 things to do here
1. `SearchDashboardVariable` should validate that `inputType` is not `input`
2. `ISearchDashboardVariable.populateFrom` should not be optional

I think `ValueDashboardVariable` might need some similar validations as well